js/views/file_input_view.js

```diff
@@ -1,236 +1,314 @@
-
+/* global Whisper: false */
+/* global i18n: false */
+/* global loadImage: false */
+/* global Backbone: false */
+/* global _: false */
+/* global Signal: false */
+
+// eslint-disable-next-line func-names
 (function () {
-    'use strict';
-    window.Whisper = window.Whisper || {};
-
-    const { MIME } = window.Signal.Types;
-
-    Whisper.FileSizeToast = Whisper.ToastView.extend({
-        templateName: 'file-size-modal',
-        render_attributes: function() {
-            return {
-                'file-size-warning': i18n('fileSizeWarning'),
-                limit: this.model.limit,
-                units: this.model.units
-            };
-        }
+  'use strict';
+
+  window.Whisper = window.Whisper || {};
+
+  const { MIME } = window.Signal.Types;
+
+  Whisper.FileSizeToast = Whisper.ToastView.extend({
+    templateName: 'file-size-modal',
+    render_attributes() {
+      return {
+        'file-size-warning': i18n('fileSizeWarning'),
+        limit: this.model.limit,
+        units: this.model.units,
+      };
+    },
+  });
+  Whisper.UnsupportedFileTypeToast = Whisper.ToastView.extend({
+    template: i18n('unsupportedFileType'),
+  });
+
+  function makeThumbnail(size, objectUrl) {
+    return new Promise(((resolve, reject) => {
+      const img = document.createElement('img');
+      img.onerror = reject;
+      img.onload = () => {
+        // using components/blueimp-load-image
+
+        // first, make the correct size
+        let canvas = loadImage.scale(img, {
+          canvas: true,
+          cover: true,
+          maxWidth: size,
+          maxHeight: size,
+          minWidth: size,
+          minHeight: size,
+        });
+
+        // then crop
+        canvas = loadImage.scale(canvas, {
+          canvas: true,
+          crop: true,
+          maxWidth: size,
+          maxHeight: size,
+          minWidth: size,
+          minHeight: size,
+        });
+
+        const blob = window.dataURLToBlobSync(canvas.toDataURL('image/png'));
+
+        resolve(blob);
+      };
+      img.src = objectUrl;
+    }));
+  }
+
+  function makeVideoScreenshot(objectUrl) {
+    return new Promise(((resolve, reject) => {
+      const video = document.createElement('video');
+
+      function capture() {
+        const canvas = document.createElement('canvas');
+        canvas.width = video.videoWidth;
+        canvas.height = video.videoHeight;
+        canvas.getContext('2d').drawImage(video, 0, 0, canvas.width, canvas.height);
+
+        const image = window.dataURLToBlobSync(canvas.toDataURL('image/png'));
+
+        video.removeEventListener('canplay', capture);
+
+        resolve(image);
+      }
+
+      video.addEventListener('canplay', capture);
+      video.addEventListener('error', (error) => {
+        console.log(
+          'makeVideoThumbnail error',
+          error && error.stack ? error.stack : error
+        );
+        reject(error);
+      });
+
+      video.src = objectUrl;
+    }));
+  }
```

Comment:
Could we add new code such as this into their own module instead of making our Backbone code bigger? I feel that’ll make it easier to move away from Backbone and reduces the chance we rely on any underlying data.

Reply:
I know you want everything moved to typescript. But I always choose the most natural way to write new code. In this case, adding it to the file where the other thumbnail logic already resides.

Reply:
It’s not necessarily about TypeScript but rather modularity. Smaller modules are easier to understand, reuse, test, and later move between different technologies, e.g. Backbone --> React. If you feel more comfortable feel free to do `js/modules` instead of TypeScript.